xds/src/test/java/io/grpc/xds/XdsSdsClientServerTest.java

```diff
@@ -0,0 +1,224 @@
+import static org.junit.Assert.fail;
+import static org.mockito.Mockito.mock;
+
+import io.envoyproxy.envoy.api.v2.auth.DownstreamTlsContext;
+import io.envoyproxy.envoy.api.v2.auth.UpstreamTlsContext;
+import io.grpc.Server;
+import io.grpc.StatusRuntimeException;
+import io.grpc.stub.StreamObserver;
+import io.grpc.testing.GrpcCleanupRule;
+import io.grpc.testing.protobuf.SimpleRequest;
+import io.grpc.testing.protobuf.SimpleResponse;
+import io.grpc.testing.protobuf.SimpleServiceGrpc;
+import io.grpc.xds.internal.sds.SdsProtocolNegotiators;
+import io.grpc.xds.internal.sds.SecretVolumeSslContextProviderTest;
+import io.grpc.xds.internal.sds.XdsChannelBuilder;
+import io.grpc.xds.internal.sds.XdsServerBuilder;
+import java.io.IOException;
+import java.net.ServerSocket;
+import java.util.Arrays;
+import javax.net.ssl.SSLHandshakeException;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/**
+ * Unit tests for {@link XdsChannelBuilder} and {@link XdsServerBuilder} for plaintext/TLS/mTLS
+ * modes.
+ */
+@RunWith(JUnit4.class)
+public class XdsSdsClientServerTest {
+
+  /** Bad/untrusted server certs. */
+  private static final String BAD_SERVER_PEM_FILE = "badserver.pem";
+  private static final String BAD_SERVER_KEY_FILE = "badserver.key";
+
+  @Rule public final GrpcCleanupRule cleanupRule = new GrpcCleanupRule();
+  private Server server;
+
+  @Test
+  public void plaintextClientServer() throws IOException {
+    getXdsServer(/* downstreamTlsContext= */ null);
+    buildClientAndTest(
+        /* upstreamTlsContext= */ null, /* overrideAuthority= */ null, "buddy", server.getPort());
+  }
+
+  /** TLS channel - no mTLS. */
+  @Test
+  public void tlsClientServer_noClientAuthentication() throws IOException {
+    DownstreamTlsContext downstreamTlsContext =
+        SecretVolumeSslContextProviderTest.buildDownstreamTlsContextFromFilenames(
+            SecretVolumeSslContextProviderTest.SERVER_1_KEY_FILE,
+            SecretVolumeSslContextProviderTest.SERVER_1_PEM_FILE,
+            null);
+
+    getXdsServer(downstreamTlsContext);
+
+    // for TLS, client only needs trustCa
+    UpstreamTlsContext upstreamTlsContext =
+        SecretVolumeSslContextProviderTest.buildUpstreamTlsContextFromFilenames(
+            null, null, SecretVolumeSslContextProviderTest.CA_PEM_FILE);
+    buildClientAndTest(upstreamTlsContext, "foo.test.google.fr", "buddy", server.getPort());
+  }
+
+  /** mTLS - client auth enabled. */
+  @Test
+  public void mtlsClientServer_withClientAuthentication() throws IOException {
+    UpstreamTlsContext upstreamTlsContext =
+        SecretVolumeSslContextProviderTest.buildUpstreamTlsContextFromFilenames(
+            SecretVolumeSslContextProviderTest.CLIENT_KEY_FILE,
+            SecretVolumeSslContextProviderTest.CLIENT_PEM_FILE,
+            SecretVolumeSslContextProviderTest.CA_PEM_FILE);
+    mtlsCommonTest(upstreamTlsContext);
```

Comment:
is this testing test utility method?

Reply:
No, this is testing a complete mTLS scenario but used more than once to test the listener update scenario - hence the util function `mtlsCommonTest`